src/Compilers/CSharp/Test/Emit3/Semantics/DictionaryExpressionTests.cs

```diff
@@ -0,0 +1,616 @@
+                  IL_0026:  callvirt   "void System.Collections.Generic.Dictionary<int, string>.Add(int, string)"
+                  IL_002b:  ldloc.0
+                  IL_002c:  ldc.i4.2
+                  IL_002d:  call       "int Program.Identity<int>(int)"
+                  IL_0032:  ldstr      "two"
+                  IL_0037:  call       "string Program.Identity<string>(string)"
+                  IL_003c:  callvirt   "void System.Collections.Generic.Dictionary<int, string>.Add(int, string)"
+                  IL_0041:  ldc.i4.1
+                  IL_0042:  newarr     "System.Collections.Generic.KeyValuePair<int, string>"
+                  IL_0047:  dup
+                  IL_0048:  ldc.i4.0
+                  IL_0049:  ldc.i4.3
+                  IL_004a:  ldstr      "three"
+                  IL_004f:  newobj     "System.Collections.Generic.KeyValuePair<int, string>..ctor(int, string)"
+                  IL_0054:  stelem     "System.Collections.Generic.KeyValuePair<int, string>"
+                  IL_0059:  call       "System.Collections.Generic.KeyValuePair<int, string>[] Program.Identity<System.Collections.Generic.KeyValuePair<int, string>[]>(System.Collections.Generic.KeyValuePair<int, string>[])"
+                  IL_005e:  stloc.3
+                  IL_005f:  ldc.i4.0
+                  IL_0060:  stloc.s    V_4
+                  IL_0062:  br.s       IL_0087
+                  IL_0064:  ldloc.3
+                  IL_0065:  ldloc.s    V_4
+                  IL_0067:  ldelem     "System.Collections.Generic.KeyValuePair<int, string>"
+                  IL_006c:  stloc.2
+                  IL_006d:  ldloc.0
+                  IL_006e:  ldloca.s   V_2
+                  IL_0070:  call       "int System.Collections.Generic.KeyValuePair<int, string>.Key.get"
+                  IL_0075:  ldloca.s   V_2
+                  IL_0077:  call       "string System.Collections.Generic.KeyValuePair<int, string>.Value.get"
+                  IL_007c:  callvirt   "void System.Collections.Generic.Dictionary<int, string>.Add(int, string)"
+                  IL_0081:  ldloc.s    V_4
+                  IL_0083:  ldc.i4.1
+                  IL_0084:  add
+                  IL_0085:  stloc.s    V_4
+                  IL_0087:  ldloc.s    V_4
+                  IL_0089:  ldloc.3
+                  IL_008a:  ldlen
+                  IL_008b:  conv.i4
+                  IL_008c:  blt.s      IL_0064
+                  IL_008e:  ldloc.0
+                  IL_008f:  call       "void DictionaryExtensions.Report<int, string>(System.Collections.Generic.IEnumerable<System.Collections.Generic.KeyValuePair<int, string>>)"
+                  IL_0094:  ret
+                }
+                """);
+        }
+
+        [Fact]
+        public void InferredType_ExpressionElement()
+        {
+            string source = """
+                using System.Collections.Generic;
+                IDictionary<string, int> d = [default, new()];
```

Comment:
What are the inferred types? Same question below (in `InferredType_KeyValueElement`)
Consider executing and reporting the contents
Also, it's not blocking for this PR, but at some point we'll want to verify the semantic model (types and conversions) on the elements in these new scenario